File: App/src/TestPerformance.cpp
// ************************************************************************** //
//
//  BornAgain: simulate and fit scattering at grazing incidence
//
//! @file      App/src/TestPerformance.cpp
//! @brief     Implements TestPerformance
//!
//! @homepage  http://www.bornagainproject.org
//! @license   GNU General Public License v3 or higher (see COPYING)
//! @copyright Forschungszentrum Jülich GmbH 2015
//! @authors   Scientific Computing Group at MLZ Garching
//! @authors   C. Durniak, M. Ganeva, G. Pospelov, W. Van Herck, J. Wuttke
//
// ************************************************************************** //

#include "TestPerformance.h"
#include "SimulationRegistry.h"
#include "SampleBuilderFactory.h"
#include "Utils.h"
#include "ProgramOptions.h"
#include "Units.h"
#include "SpecularMatrix.h"
#include "SampleFactory.h"
#include "SpecularMagnetic.h"

#include "TDatime.h"
#include "TSystem.h"
#include "TBenchmark.h"
#include <sstream>
#include <fstream>
#include <iomanip>
#include <boost/format.hpp>



// -----------------------------------------------------------------------

TestPerformance::TestPerformance()
{
    add("MultiLayer",         "MaxiGISAS",    "MultiLayerWithRoughnessBuilder", 1);
    add("CylindersInDWBA",    "MaxiGISAS",    "CylindersInDWBABuilder", 10);
    add("RotatedPyramids",    "MaxiGISAS",    "RotatedPyramidsBuilder", 10);
    add("CoreShell",          "MaxiGISAS",    "CoreShellParticleBuilder", 10);
    add("SquareLattice",      "MaxiGISAS",    "SquareLatticeBuilder", 10);
    add("RadialParaCrystal",  "MaxiGISAS",    "RadialParaCrystalBuilder", 10);
    add("HexParaCrystal",     "BasicGISAS",   "HexParaCrystalBuilder", 1);
    add("SSCA",               "MaxiGISAS",    "SizeDistributionSSCAModelBuilder", 10);
    add("Mesocrystal",        "MaxiGISAS",    "MesoCrystalBuilder", 2);
    add("PolMagCyl",          "MaxiGISAS00",  "MagneticCylindersBuilder", 10);
    std::cout << "TestPerformance::TestPerformance() -> Info. Preparing to run "
              << m_tests.size() << " performance tests."  << std::endl;
}


TestPerformance::~TestPerformance()
{
}


void TestPerformance::add(const std::string &name, const std::string &simulation_name,
                          const std::string &sample_builder, size_t nrepetitions)
{
    m_tests.push_back(PerfTest(name, simulation_name, sample_builder, nrepetitions));
}




void TestPerformance::execute()
{

    init_sysinfo();
    for(size_t i=0; i<m_tests.size(); ++i) {
        PerfTest *test = &m_tests[i];
        runTest(test);
    }

    write_results();
}


void TestPerformance::runTest(PerfTest *test)
{
    SimulationRegistry simulationRegistry;
    SampleBuilderFactory sampleFactory;

    std::cout << "Running test: " << std::setw(20) << std::left << test->m_test_name << " ... ";
    std::cout.flush();

    boost::scoped_ptr<Simulation> simulation(simulationRegistry.createSimulation(test->m_simulation_name));
    SampleBuilder_t sample = sampleFactory.createBuilder(test->m_sample_builder_name);

    TBenchmark mb;
    mb.Start("test");

    for(size_t i=0; i<test->m_nrepetitions; ++i) {
        simulation->setSampleBuilder(sample);
        simulation->runSimulation();
    }

    mb.Stop("test");
    test->m_cpu_time = mb.GetCpuTime("test");
    test->m_real_time = mb.GetRealTime("test");

    std::cout << boost::format("OK: %-6.3f (real sec), %-6.3f (cpu sec) ") % test->m_real_time % test->m_cpu_time << std::endl;
}


void TestPerformance::write_results()
{
    std::string filename("perf_history.txt");
    std::ofstream file;
    file.open(filename.c_str(), std::ios_base::app);
    if( !file.is_open() ) {
        throw FileNotIsOpenException("TestPerformance::execute() -> Error. Can't open file '"+filename+"' for writing.");
    }

    write_header(file);
    write_performance(file);

    file.close();

    std::cout << "TestPerformance::write_performance() -> Info. File '" << filename << "' is updated." << std::endl;
}


//! write header in file
void TestPerformance::write_header(std::ofstream &file)
{
    file << std::endl;
    file << boost::format("| %-19s | %-10s | %-13s | %2s | %-8s |") %  "date " % "hostname" % "sysinfo" % "tr" % "total";
    for(size_t i=0; i<m_tests.size(); ++i) {
        file  << boost::format(" %-12s |") %  Utils::AdjustStringLength(m_tests[i].m_test_name, 12);
    }
    file << std::endl;
}

//! write results of performance measurements
void TestPerformance::write_performance(std::ofstream &file)
{
    file << boost::format("| %-19s | %-10s | %-13s | %-2d |") %  m_datime % m_hostname % m_sysinfo % 0;
    double sum_real(0), sum_cpu(0);
    for(size_t i=0; i<m_tests.size(); ++i) {
        sum_real += m_tests[i].m_real_time;
        sum_cpu += m_tests[i].m_cpu_time;
    }
    file  << boost::format(" %-8.3f |") %  sum_real;

    for(size_t i=0; i<m_tests.size(); ++i) {
        file  << boost::format(" %-12.3f |") %  m_tests[i].m_real_time;
    }
    file << std::endl;
}


//! read sustem information and store it in the test
void TestPerformance::init_sysinfo()
{
    // saving date and time
    TDatime td;
    m_datime = std::string(td.AsSQLString());

    // saving host name
    std::string hostname(gSystem->HostName());
    // stripping host name after first '.' (somehost.jcns.frm2 -> somehost)
    std::string::size_type pos = hostname.find_first_of('.');
    if(pos != std::string::npos) {
        hostname.erase(pos,hostname.size()-pos);
    }
    m_hostname = hostname;

    // saving hardware information
    SysInfo_t sys_info;
    int status = gSystem->GetSysInfo(&sys_info);
    if( status == -1) {
        m_sysinfo = std::string("Unknown hardware");
    }else{
        std::ostringstream os;
        os << std::string(gSystem->GetBuildArch());
        //os << std::string(gSystem->GetBuildArch()) << ", "<< sys_info.fCpuSpeed << " MHz";
        //os << ", " << 	sys_info.fL2Cache << " Kb";
        m_sysinfo = os.str();
    }
}


// -----------------------------------------------------------------------------
// General PerformanceTest
// -----------------------------------------------------------------------------

//! run performance measurements
//void PerformanceTest::execute()
//{
//    TBenchmark mb;
//    mb.Start( getName().c_str() );

//    runTests();

//    mb.Stop( getName().c_str() );

//    m_cpu_time = mb.GetCpuTime(getName().c_str());
//    m_real_time = mb.GetRealTime(getName().c_str());
//}


////! run standard simulation
//void PerformanceTest::runTests()
//{
//    SimulationRegistry registry;
//    Simulation *simulation = registry.createSimulation(getName());
//    simulation->setProgramOptions(mp_options);
//    for(int i=0; i<m_nrepetitions; i++){
//        simulation->runSimulation();
//    }
//    delete simulation;
//}

//// -----------------------------------------------------------------------------
//// custom PerformanceTest's
//// -----------------------------------------------------------------------------

//void SpecularMatrixPerformanceTest::runTests()
//{
//    MultiLayer *ml = dynamic_cast<MultiLayer *>(SampleFactory::createSample("SimpleMultilayer"));

//    for(int i=0; i<m_nrepetitions; i++){
//        static double alpha_i = -0.3;
//        kvector_t kvec;
//        kvec.setLambdaAlphaPhi(1.54*Units::angstrom, -alpha_i, 0.0);
//        SpecularMatrix::MultiLayerCoeff_t coeffs;
//        SpecularMatrix matrixCalculator;
//        matrixCalculator.execute(*ml, kvec, coeffs);
//    }
//    delete ml;
//}

//void SpecularMagneticPerformanceTest::runTests()
//{
//    MultiLayer *ml = dynamic_cast<MultiLayer *>(SampleFactory::createSample("MultilayerSpecularMagneticTestCase"));

//    for(int i=0; i<m_nrepetitions; i++){
//        static double alpha_i = -0.3;
//        kvector_t kvec;
//        kvec.setLambdaAlphaPhi(1.54*Units::angstrom, -alpha_i, 0.0);
//        SpecularMagnetic::MultiLayerCoeff_t coeffs;
//        SpecularMagnetic magneticCalculator;
//        magneticCalculator.execute(*ml, kvec, coeffs);
//    }
//    delete ml;
//}

File: Tests/PerformanceTests/test_performance.py
#!/usr/bin/python
 # -*- coding: utf-8 -*-

## ************************************************************************** ##
##
##  BornAgain: simulate and fit scattering at grazing incidence
##
##! @file      test_performance.py
##! @brief     Defines class TestPerformance for logging performance changes
##!            For explanations, see the README file
##!
##! @homepage  http:##www.bornagainproject.org
##! @license   GNU General Public License v3 or higher (see COPYING)
##! @copyright Forschungszentrum Jülich GmbH 2016
##! @authors   Scientific Computing Group at MLZ Garching
##! @authors   C. Durniak, M. Ganeva, G. Pospelov, W. Van Herck, J. Wuttke
##
## ************************************************************************** ##

from __future__ import print_function
import datetime
import sys
import platform
import timeit
from collections import OrderedDict
import math
from libBornAgainCore import *

# used for logging intermediate output
logfile = sys.stdout

# for code readability
get_wall_time = timeit.default_timer

# for code readability and portability, since resource is not guaranteed to be available
try:
    import resource
    record_cpu_time = True
    get_cpu_time = lambda: resource.getrusage(resource.RUSAGE_SELF)[0]
except:
    record_cpu_time = False
    get_cpu_time = lambda: None


# globals used in custom form factor
phi_min, phi_max = -1.0, 1.0
alpha_min, alpha_max = 0.0, 2.0

# user-defined custom form factor
class CustomFormFactor(IFormFactorBorn):
    """
    A custom defined form factor
    The form factor is V sech(q L) with
    V volume of particle
    L length scale which defines mean radius
    """
    def __init__(self, V, L):
        IFormFactorBorn.__init__(self)
        # parameters describing the form factor
        self.V = V
        self.L = L

    def clone(self):
        """
        IMPORTANT NOTE:
        The clone method needs to call transferToCPP() on the cloned object
        to transfer the ownership of the clone to the cpp code
        """
        cloned_ff = CustomFormFactor(self.V, self.L)
        cloned_ff.transferToCPP()
        return cloned_ff

    def evaluate_for_q(self, q):
        return self.V*1.0/math.cosh(q.mag()*self.L)

# class for performance test, constructed using sample factories
class FactoryTest:
    def __init__(self, name, simulation_name, sample_builder, nrepetitions):
        self.m_test_name = name
        self.m_simulation_name = simulation_name
        self.m_sample_builder_name = sample_builder
        self.m_nrepetitions = nrepetitions
        self.m_cpu_time = 0.0
        self.m_wall_time = 0.0

        self.m_sample = None

        if simulation_name != None and sample_builder != None:
            self.m_sample_factory = SampleBuilderFactory()
            self.m_simulation_factory = SimulationFactory()
            self.m_simulation = self.m_simulation_factory.createItem(self.m_simulation_name)
            self.m_sample = self.m_sample_factory.createSample(self.m_sample_builder_name)
        else:
            self.m_sample_factory = None
            self.m_simulation_factory = None
            self.m_simulation = None
            self.m_sample = None

    def prepare(self):
        pass

    # do the actual work
    def run_loop(self):
        # actual work is done here
        for i in range(self.m_nrepetitions):
            self.m_simulation.setSample(self.m_sample)
            self.m_simulation.runSimulation()

    def run(self):

        self.prepare()

        logfile.write("Running test: %-30s " % self.m_test_name)
        logfile.flush()

        # resource.getrusage is less accurate but returns user (cpu) time
        start_wall_time = get_wall_time()
        start_cpu_time = get_cpu_time()

        self.run_loop()

        end_wall_time = get_wall_time()
        end_cpu_time = get_cpu_time()

        # note that on multi-core system, cpu time can be higher than wall time
        self.m_wall_time = end_wall_time - start_wall_time

        if record_cpu_time:
            self.m_cpu_time = end_cpu_time - start_cpu_time
            logfile.write("OK: %-6.3f (wall sec), %-6.3f (cpu sec) \n" % (self.m_wall_time, self.m_cpu_time))
        else:
            logfile.write("OK: %-6.3f (wall sec)\n" % (self.m_wall_time))



# special performance test case: custom form factor
class CustomTest(FactoryTest):
    def __init__(self, name, nrepetitions):
        FactoryTest.__init__(self, name, None, None, nrepetitions)
        self.m_sample = None
        self.m_simulation = None

    def prepare(self):
        self.m_sample = self.get_sample()
        self.m_simulation = self.get_simulation()

    # do the actual work
    def run_loop(self):
        for i in range(self.m_nrepetitions):
            self.m_simulation.setSample(self.m_sample)
            self.m_simulation.runSimulation()

    def get_sample(self):
        """
        Build and return the sample to calculate custom form factor in Distorted Wave Born Approximation.
        """
        # defining materials
        m_ambience = HomogeneousMaterial("Air", 0.0, 0.0)
        m_substrate = HomogeneousMaterial("Substrate", 6e-6, 2e-8)
        m_particle = HomogeneousMaterial("Particle", 6e-4, 2e-8)

        # collection of particles
        ff = CustomFormFactor(343.0*nanometer, 7.0*nanometer)
        particle = Particle(m_particle, ff)
        particle_layout = ParticleLayout()
        particle_layout.addParticle(particle, 1.0)
        air_layer = Layer(m_ambience)
        air_layer.addLayout(particle_layout)
        substrate_layer = Layer(m_substrate)

        # assemble multilayer
        multi_layer = MultiLayer()
        multi_layer.addLayer(air_layer)
        multi_layer.addLayer(substrate_layer)
        return multi_layer

    def get_simulation(self):
        """
        Create and return GISAXS simulation with beam and detector defined
        IMPORTANT NOTE:
        Multithreading should be deactivated by putting ThreadInfo.n_threads to -1
        """
        simulation = GISASSimulation()
        simulation.getOptions().setNumberOfThreads(-1)
        simulation.setDetectorParameters(100, phi_min*degree, phi_max*degree, 100, alpha_min*degree, alpha_max*degree)
        simulation.setBeamParameters(1.0*angstrom, 0.2*degree, 0.0*degree)
        return simulation


# class for running all of the tests and collecting results
class PerformanceTests:
    def __init__(self, filename=None):
        self.m_tests = []
        self.m_datime = ""
        self.m_hostname = ""
        self.m_sysinfo = ""
        self.m_pyversion = ""
        self.m_filename = filename

        self.add("MultiLayer",         "MaxiGISAS",    "MultiLayerWithRoughnessBuilder", 1)
        self.add("CylindersInDWBA",    "MaxiGISAS",    "CylindersInDWBABuilder", 10)
        self.add("RotatedPyramids",    "MaxiGISAS",    "RotatedPyramidsBuilder", 10)
        self.add("CoreShell",          "MaxiGISAS",    "CoreShellParticleBuilder", 10)
        self.add("SquareLattice",      "MaxiGISAS",    "SquareLatticeBuilder", 10)
        self.add("RadialParaCrystal",  "MaxiGISAS",    "RadialParaCrystalBuilder", 10)
        self.add("HexParaCrystal",     "BasicGISAS",   "HexParaCrystalBuilder", 1)
        self.add("SSCA",               "MaxiGISAS",    "SizeDistributionSSCAModelBuilder", 10)
        self.add("Mesocrystal",        "MaxiGISAS",    "MesoCrystalBuilder", 2)
        self.add("PolMagCyl",          "MaxiGISAS00",  "MagneticCylindersBuilder", 10)

        # custom form factor is a special case since it's not in the registry
        self.m_tests.append(CustomTest("Custom FF", 10))

        logfile.write("\nPreparing to run %d performance tests.\n\n" % len(self.m_tests))

    def add(self, name, simulation_name, sample_builder, nrepetitions):
        self.m_tests.append(FactoryTest(name, simulation_name, sample_builder, nrepetitions))

    # execute all performance tests
    def execute(self):
        self.init_sysinfo()
        for test in self.m_tests: test.run()
        self.write_results()


    # write out system information and test results to file
    def write_results(self):

        if ( self.m_filename != None ):
            try:
                write_file = open(self.m_filename, "a")
            except:
                sys.stderr.write("Could not open filed '%' for writing. Writing to stdout instead.\n" % self.m_filename)
                write_file = sys.stdout
                self.m_filename = None

        else:
            write_file = sys.stdout

        logfile.write("\nWriting output to %s...\n\n" % write_file.name)

        # record results into an ordered dict, which is used by pretty_write() below
        dictionary = OrderedDict()

        dictionary["date"] = self.m_datime
        dictionary["hostname"] = self.m_hostname
        dictionary["sysinfo"] = self.m_sysinfo
        dictionary["python"] = self.m_pyversion

        sum_wall = 0.0
        sum_cpu = 0.0

        for test in self.m_tests:
            sum_wall += test.m_wall_time
            sum_cpu += test.m_cpu_time

        if record_cpu_time: dictionary["total cpu"] = "%-.4f" % sum_cpu
        dictionary["total wall"] = "%-.4f" % sum_wall

        for test in self.m_tests:
            dictionary[test.m_test_name] = "%-.4f" % test.m_wall_time

        write_file.write("\n")
        pretty_write(dictionary, write_file)
        write_file.write("\n")
        write_file.flush()

        if ( self.m_filename != None ):
            write_file.close()

    # determine platform, architecture, python version, etc.
    def init_sysinfo(self):
        system, node, release, version, machine, processor = platform.uname()
        self.m_datime = datetime.datetime.strftime(datetime.datetime.now(), '%Y-%m-%d %H:%M:%S')
        self.m_hostname = node
        self.m_sysinfo = "%s %s" % (system, machine)
        self.m_pyversion = "%d.%d" % (sys.version_info[:2])

# used for writing the summary with proper formatting
def pretty_write(dictionary, file):
    header = "|"
    footer = "|"

    for key in dictionary:
        value = dictionary[key]
        length = max(len(key), len(value))
        format_str = " {0:" + str(length) + "} |"
        header += format_str.format(key)
        footer += format_str.format(value)

    file.write(header + "\n")
    file.write(footer + "\n")

if __name__ == '__main__':

    if ( len(sys.argv) > 2 ):
        print("Usage: test_performance.py [output file]")
        exit(0)

    if (len(sys.argv) == 2):
        filename = sys.argv[1]
    else:
        filename = None

    tests = PerformanceTests(filename)
    tests.execute()
